src/sagemaker_containers/_intermediate_output.py

```diff
@@ -0,0 +1,165 @@
+        for event in inotify.read(timeout=1000):
+            for flag in flags.from_mask(event.mask):
+                # if new directory was created traverse the directory tree to recursively add all
+                # created folders to the watchers list.
+                # Upload files to s3 if there any files.
+                # There is a potential race condition if upload the file and the see a notification
+                # for it which should cause any problems because when we copy files to temp dir
+                # we add a unique timestamp up to microseconds.
+                if flag is flags.ISDIR:
+                    for folder, dirs, files in os.walk(os.path.join(intermediate_path, event.name)):
+                        wd = inotify.add_watch(folder, watch_flags)
+                        relative_path = os.path.relpath(folder, intermediate_path)
+                        watchers[wd] = relative_path
+                        os.makedirs(os.path.join(tmp_dir_path, relative_path))
+                        for file in files:
+                            _copy_file(executor, s3_uploader, relative_path, file)
+                elif flag is flags.CLOSE_WRITE:
+                    _copy_file(executor, s3_uploader, watchers[event.wd], event.name)
+
+        last_pass_done = stop_file_exists
+        stop_file_exists = os.path.exists(success_file_path) or os.path.exists(failure_file_path)
+
+    # wait for all the s3 upload tasks to finish and shutdown the executor
+    executor.shutdown(wait=True)
+
+
+def start_intermediate_folder_sync(s3_output_location, region):
+    """Starts intermediate folder sync which copies files from 'opt/ml/output/intermediate'
+    directory to the provided s3 output location as files created or modified.
+    If files are deleted it doesn't delete them from s3.
+
+    It starts intermediate folder behavior as a daemonic process and
+    only if the directory doesn't exists yet, if it does - it indicates
+    that platform is taking care of syncing files to S3 and container should not interfere.
+
+    Args:
+        s3_output_location (str): name of the script or module.
+        region (str): the location of the module.
+
+    Returns:
+        (Process): the intermediate output sync daemonic process.
+    """
+    if not s3_output_location or os.path.exists(intermediate_path):
+        logger.debug('Could not initialize intermediate folder sync to s3.')
+        return None
+
+    # create intermediate and intermediate_tmp directories
+    os.makedirs(intermediate_path)
+    os.makedirs(tmp_dir_path)
+
+    # configure unique s3 output location similar to how SageMaker platform does it
+    # or link it to the local output directory
+    url = urlparse(s3_output_location)
+    if url.scheme == 'file':
+        logger.debug('Local directory is used for output. No need to sync any intermediate output.')
+        return
```

Comment:
nit: I personally avoid returning in the middle of the fn like that, it makes easier to have logic errors

Reply:
+1 this is a guard and doesn't depend on the preceding ops so move it up

Reply:
ok. side effect before the check is wanted even if sync is disabled